modules/navier_stokes/src/fvkernels/INSFVAdvectionKernel.C

```diff
@@ -43,27 +34,38 @@ INSFVAdvectionKernel::INSFVAdvectionKernel(const InputParameters & params)
              "configure script in the root MOOSE directory with the configure option "
              "'--with-ad-indexing-type=global'");
 #endif
-  using namespace Moose::FV;
+  const bool need_more_ghosting =
+      Moose::FV::setInterpolationMethods(*this, _advected_interp_method, _velocity_interp_method);
+  if (need_more_ghosting && _tid == 0)
+  {
+    auto & factory = _app.getFactory();
+
+    auto rm_params = factory.getValidParams("ElementSideNeighborLayers");
+
+    rm_params.set<std::string>("for_whom") = name();
+    rm_params.set<MooseMesh *>("mesh") = &const_cast<MooseMesh &>(_mesh);
+    rm_params.set<Moose::RelationshipManagerType>("rm_type") =
+        Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+        Moose::RelationshipManagerType::COUPLING;
+    FVKernel::setRMParams(
+        _pars, rm_params, std::max((unsigned short)(3), _pars.get<unsigned short>("ghost_layers")));
+    mooseAssert(rm_params.areAllRequiredParamsValid(),
+                "All relationship manager parameters should be valid.");
+
+    auto rm_obj = factory.create<RelationshipManager>(
+        "ElementSideNeighborLayers", name() + "_skew_correction", rm_params);
 
-  const auto & advected_interp_method = getParam<MooseEnum>("advected_interp_method");
-  if (advected_interp_method == "average")
-    _advected_interp_method = InterpMethod::Average;
-  else if (advected_interp_method == "skewness-corrected")
-    _advected_interp_method = Moose::FV::InterpMethod::SkewCorrectedAverage;
-  else if (advected_interp_method == "upwind")
-    _advected_interp_method = InterpMethod::Upwind;
-  else
-    mooseError("Unrecognized interpolation type ",
-               static_cast<std::string>(advected_interp_method));
+    // Delete the resources created on behalf of the RM if it ends up not being added to the
+    // App.
+    if (!_app.addRelationshipManager(rm_obj))
+      factory.releaseSharedObjects(*rm_obj);
 
-  const auto & velocity_interp_method = getParam<MooseEnum>("velocity_interp_method");
-  if (velocity_interp_method == "average")
-    _velocity_interp_method = InterpMethod::Average;
-  else if (velocity_interp_method == "rc")
-    _velocity_interp_method = InterpMethod::RhieChow;
-  else
-    mooseError("Unrecognized interpolation type ",
-               static_cast<std::string>(velocity_interp_method));
+    // If we need more ghosting, then we are a second-order nonlinear limiting scheme whose stencil
+    // is liable to change upon wind-direction change. Consequently we need to tell our problem that
+    // it's ok to have new nonzeros which may crop-up after PETSc has shrunk the matrix memory
```

Comment:
Is it possible to set the ghosting directly based on the interpolation/limiter types? Wouldn't this allocation slow down the whole simulation a lot? I remember running into these with the skewness correction and even simple problems essentially froze because of this.

Reply:
In the code right above we add more ghosting based on the limiter type. But that's often not good enough. You can imagine that during the first Jacobian evaluation the velocity is in the + direction. So for a second-order method, we may use cells --, -, and +. When it's done assembling, PETSc will note that our sparsity pattern is --, -, and + and it will remove the allocation for ++ even if we had provided it up front (which we do based on the code above your comment here). However, after the first nonlinear iteration and Newton update the velocity may be in the - direction and now we want cells -, +, and ++. Since we removed ++ after the first assembly, we have to reallocate it.

We have plans in the future to prevent PETSc from shrinking memory, but that has not yet been implemented.

Reply:
Okay, I see. Is there a way to throw a warning instead of an error if petsc is reallocating? So users know why the simulation slowed down.

Reply:
I not think so. @fdkong would know better than I

Reply:
For this use case, I think we should hook up coupling ghosting functors that understand what need to couple when calling different unwinding schemes , and then recreate matrix for every nonlinear iteration.

Reply:
Yeah, right now we do not have a warning.